Extend eQ-3 Bluetooth Smart thermostats integration

clau-bucur opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Extended the support for eQ-3 Bluetooth thermostats, and improved the logging.
Added the following new configuration options:
- "**target_sensor**" allows specifying an external temperature sensor to report the current temperature in the room, since these devices do not contain a temperature sensor;
- "**control_by_target_sensor**" allows to run the thermostat in manual mode controlled by the readings of the "target_sensor".

Added the following options, similar with the ones from "Generic Thermostat" integration:
- "min_temp"
- "max_temp"
- "precision"
- "cold_tolerance"
- "hot_tolerance"

When the **control_by_target_sensor** is enabled, the thermostat will run in manual mode and no presets will be available.
The only available HVAC modes will be limited to **HEAT** or **OFF**.
When mode is **HEAT**:
- When the **target temperature** is above the temperature reported by the **target_sensor**, the bluetooth thermostat will switch to "ON" mode, and the climate entity will switch to **heating**.
- When the **target temperature** is below the temperature reported by the **target_sensor**, the bluetooth thermostat will switch to "OFF" mode, and the climate entity will switch to **idle**.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
climate:
  - platform: eq3btsmart
    scan_interval: 300 #set to 5 min, to reduce the number of times the thermostats are querried
    devices:
      office:
        mac: 'xx:xx:xx:xx:xx:xx'
        target_sensor: sensor.mi_temperature_office
        control_by_target_sensor: true
        min_temp: 18
        max_temp: 24
        precision: 0.1
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

clau-bucur committed:
eq3: add min_temp, max_temp, precision

clau-bucur committed:
eq3: add external temperature sensor support

Add the following new configuration options:

- "target_sensor" allows specifying an external temperature sensor to
   report the current temperature in the room, since these devices do
   not contain a temperature sensor;
- "control_by_target_sensor" allows to run the thermostat in manual mode
   controlled by the readings of the "target_sensor".

Add the following from "Generic Thermostat" integration:
- "cold_tolerance"
- "hot_tolerance"

clau-bucur committed:
eq3: sort eq3 states according to their logical order

reaper7 commented:
very useful !

andrewshadura commented:
Hi @clau-bucur, thanks for the improvements; I’m curious, do your async changes help with reliability at least a bit? I’m running HA with two thermostats and it fails to keep the state up to date most of the time here.

clau-bucur commented:
@andrewshadura I'm running it for a month now with 3 thermostats, and it's rock solid. I've not had any issues with it, running in HassIO.
You should ensure also that your HA bluetooth device is not creating the problems.

It'd be helpful if you test it yourself also, at least we can see if it improves your stability.
Grab it from https://github.com/clau-bucur/core/blob/dev/homeassistant/components/eq3btsmart/climate.py and copy it under `HA Config / custom_components / eq3btsmart`.

See the above configuration example if you want to use the new features.

reaper7 commented:
@clau-bucur - works for me too :)
but I had to disable control_by_target_sensor (initially it seemed like a fantastic solution)
because when the temperature of the remote sensor reports that the room is warm enough
the thermostat goes into off mode turning the mechanical head to the minimum 0%.
So...form me it's a bettery killer and and most importantly the killer of the mechanical part of the head.

For me, it would be more appropriate to leave the HEAT mode but with the temperature reduced to some level in relation to target sensor...

clau-bucur commented:
@reaper7 
So what you're saying is that by turning the devices to ON and OFF again for a few times a day can kill the mechanical part of the head? Do you have some more info or data on that?

I've just started using these types of devices, so not much experience with them. Took advantage of the functionality they had exposed, but did not think about the potential damages. (if the batteries last for a winter season I declare myself happy, but we'll have to see about that).
For me, it made (and still makes) sense to turn on the thermostat while the room temperature requires it, otherwise turn it off so that the radiator does not generate heat at all.

If you elaborate a bit more on your desired working mode, I'd be happy to find a way to accomodate.

andrewshadura commented:
> @andrewshadura I'm running it for a month now with 3 thermostats, and it's rock solid. I've not had any issues with it, running in HassIO.
> You should ensure also that your HA bluetooth device is not creating the problems.

Well, I tried different BT adapters, they work differently, but never reliably.

> It'd be helpful if you test it yourself also, at least we can see if it improves your stability.
> Grab it from https://github.com/clau-bucur/core/blob/dev/homeassistant/components/eq3btsmart/climate.py and copy it under `HA Config / custom_components / eq3btsmart`. Restart HA to load the new component.

I tried this but apparently the built-in component always overrides the one I drop into `custom_components`. Or maybe I’m doing something wrong.

clau-bucur commented:
> I tried this but apparently the built-in component always overrides the one I drop into `custom_components`. Or maybe I’m doing something wrong.

Actually you should copy all three files from https://github.com/clau-bucur/core/tree/dev/homeassistant/components/eq3btsmart
That should be it, nothing else needed.

andrewshadura commented:
> Actually you should copy all three files from https://github.com/clau-bucur/core/tree/dev/homeassistant/components/eq3btsmart
> That should be it, nothing else needed.

That’s what I did, and it still took the built-in one.

clau-bucur commented:
Not sure what to tell you.
Here's mine, by accessing it via the Samba Share add-on:
![image](https://user-images.githubusercontent.com/47672949/97878915-efdb7880-1d27-11eb-9bbb-0a525468a9ac.png)

andrewshadura commented:
After I fiddled with it for a bit, it finally started. No idea what was wrong — maybe a typo somewhere.

reaper7 commented:
@clau-bucur - battery in the auto mode that I have used so far works for a year and a little...
In auto mode, the head stroke is small...but closing the head to 0% is a completely different story.

Besides, in the original application, the schedule also sets the minimum temperature and you only specify at what times it should be higher.

I can see you know this code quite well...
so...if you can find some time, could you prepare a version with an additional boolean parameter - I don't want to come up with a name right now - but something like "go to min_temp instead off"
and change the code so that instead of turning off the head completely leave the heat mode with the set temperature declared as min_temp (if boolean parameter is set)...
it is possible?

perhaps, this additional parameter you define maybe should be also a numeric value? where zero mean current solution (OFF)
and the values above are the value to which to set if it already heats the room to the temp specified by target

clau-bucur commented:
Got it.

What about letting you override both the OFF and ON temperatures with two optional float parameters?
That way, you can define both low and high temperature limits the thermostat will go to.

reaper7 commented:
sounds good! the thermostat will operate in a narrower range.
i didn't think about it, i just wrote about observations after a few days of use

clau-bucur commented:
Done.
New options:
```
off_temp: the temperature that the thermostat will be set to when it should not be heating.
on_temp: the temperature that the thermostat will be set to when it should be heating.
```
_When these new options are used, the thermostat will work in Manual mode._

@reaper7 Please run tests and let me know how does it work.

reaper7 commented:
tnx!
Test in progress... I don't know if I can test it well until the weekend.
Best time for testing is the weekend because I have a valve practically near my ear :D

rytilahti reviewed (commented):
It's delightful to see that someone is giving some love to this integration, thanks @clau-bucur! :+1: Alas, I don't currently have time to do a proper review, nor am I using these devices at all, so I wanted to ask if you'd be willing to maintain this (and add yourself to the CODEMAINTAINERS)?

reaper7 commented:
@clau-bucur,  observations
my settings:
```yaml
- platform: eq3btsmart
  scan_interval: 300
  devices:
    salon_valve:
      mac: '00:1A:22:xx:xx:xx'
      target_sensor: sensor.mibt_salon_t
      control_by_target_sensor: true
      min_temp: 19.5
      max_temp: 24
      off_temp: 19.5
      on_temp: 24
      precision: 0.1
```
![valve1](https://user-images.githubusercontent.com/4134851/97969357-3b416580-1dc0-11eb-9b35-1ed2e4152812.jpg)

as You see - MODE is set to HEAT, temperature on the thermostat is set to 19.5, target sensor reports real temperature as 21.8,
integration shows STATE as IDLE but thermostat has an open valve at 7%.
so, in this case, in my opinion, if valve is open (>0%) then STATE shouldn't be reported as IDLE but HEATING...

clau-bucur commented:
> 
> It's delightful to see that someone is giving some love to this integration, thanks @clau-bucur! 👍 Alas, I don't currently have time to do a proper review, nor am I using these devices at all, so I wanted to ask if you'd be willing to maintain this (and add yourself to the CODEMAINTAINERS)?

My pleasure. Thank you for the integration.
I am willing to maintain it, sure! As long as I have these devices it would not be a problem.
I'm just starting with Python so I'm not experienced at all, I hope I can still be of help.

>
>  so, in this case, in my opinion, if valve is open (>0%) then STATE shouldn't be reported as IDLE but HEATING...

I understand what you're saying, and it makes some sense. But from how I see it now, it can become a bit counterintuitive.
I mean, your `hvac mode` is _HEAT_, your `target_temperature` is lower than your `current_temperature` but still your device is _HEATING_.
The role of your thermostat would be to stop heating once it reached the target temperature, no?

I propose the following:
- keep this PR as it is for now, and let's hope it gets into the core;
- let's move the discussion to Gitter on [eq3-thermostats](https://gitter.im/eq3-thermostats/community?utm_source=share-link&utm_medium=link&utm_campaign=share-link). I've created this room so we can talk on how to further improve this integration and not to spam this PR.
- once we have a working solution I'll create a new PR with the upgraded functionality.

clau-bucur commented:
@rytilahti Do I need to do anything else for this PR to get approved?

rytilahti commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -181,18 +474,53 @@ def preset_modes(self):
+            _LOGGER.error("%s - Invalid turn on request", self._name)
+            return
+
+        _LOGGER.info("%s - Turning on thermostat", self._name)
```
```suggestion
        _LOGGER.debug("%s - Turning on thermostat", self._name)
```
I think the consensus is not to log regular happenings at info level.

rytilahti replied:
Just a note, there were also other places where this was done incorrectly, I just didn't want to pollute the review too much :-)

rytilahti commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -106,52 +343,106 @@ def temperature_unit(self):
+        if self.hvac_mode == HVAC_MODE_OFF:
+            return CURRENT_HVAC_OFF
+
+        return EQ_TO_HA_HVAC_ACTION.get(self._thermostat.mode, "-")
```
The docstring indicates that the return value needs to be `CURRENT_HVAC_*` but this falls back to `-` on unknown modes.

clau-bucur replied:
I wanted to show some invalid state in case the thermostat is not in the expected state, and could not find a suitable value in CURRENT_HVAC_*.
So I suspect this is not ok. What would be the correct approach ?

rytilahti replied:
Sorry, I don't know about that, I did just briefly checked parts of the diff and saw that there was a discrepancy between the docstring and the code :-(

rytilahti commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -20,79 +24,312 @@
+        _max_temp,
+        _precision,
+        _cold_tolerance,
+        _hot_tolerance,
```
The parameters should not be underscore prefixed as that is reserved for internals. A question, would it make more sense to pass the config and extract the data inside `__init__` instead of having a monster parameter listing? :-)

clau-bucur replied:
I will rename the parameters.

I can't answer the question though, not that experienced here.
Let me know if I shall change the approach.

rytilahti commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -20,79 +24,312 @@
+
+    async def async_added_to_hass(self):
+        """Run when entity about to be added."""
+        await super().async_added_to_hass()
```
Is this necessary?

clau-bucur replied:
Used the same approach as [generic_thermostat](https://github.com/home-assistant/core/blob/b2c9bd2ca67df277093790e3fc0def34d9f7fb04/homeassistant/components/generic_thermostat/climate.py#L186).
If it's there, I assumed it's needed.

rytilahti reviewed (commented):
I went briefly through parts of the code and added a couple of minor comments. As I'm not using these nor do I have knowledge about how climate platforms are expected to behave, I feel that I'm not the correct person to make the call to merge, sorry! I added the second-opinion-wanted label and I hope someone with knowledge on climate platforms will chime in.

clau-bucur committed:
eq3: add temperature offsets

- "off_temp_offset": used when the device needs to be turned off.
   Offset the target temperature by this value and use that when
   the device needs to be in "OFF/IDLE" state
- "on_temp_offset": used when the device needs to be heating.
   Offset the target temperature by this value and use that when
   the device needs to be in "HEATING" state

clau-bucur committed:
eq3: add unique_id support

If not specified, unique_id defaults to "<entity_id>_<MAC>".

clau-bucur committed:
eq3: add friendly_name support

clau-bucur committed:
eq3: add attributes for better error handling

- "error": indicates the error encountered when the device was last
   queried. It's None if last query was successful.
- "last_error_at": indicates the date and time the last error occurred.

clau-bucur committed:
eq3: add reload integration support

clau-bucur committed:
eq3: support setting "HVAC_MODE" with "set_temperature"

andrewshadura committed:
eq3: use a constant instead of 0 to detect availability

eq3.Mode.Unknown may change in value, so it’s better to not hardcode a
magic value

andrewshadura commented:
As mentioned in a Gitter chat, I’ve combed through these commits, split, merged, refined and rebased them: https://github.com/andrewshadura/home-assistant-core/tree/pr-42475
Please reset your branch to those commits if you’re happy with the proposed changes.

homeassistant commented:
Hi @andrewshadura, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#42475).

Once you do that we will be able to review and accept this pull request.

Thanks!

andrewshadura commented:
Cc @rytilahti, @balloob, @bimbar since you looked at this code in the past — it would be great if you could review these changes.
Thanks in advance!

MartinHjelmare commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -17,141 +28,441 @@
+    eq3.Mode.Closed: CURRENT_HVAC_IDLE,
+}
+
+DEVICE_SCHEMA = vol.Schema(
```
We no longer accept additions or changes to the yaml config of integrations. See: https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md

This integration needs to be refactored to use a config flow and config entries.

https://developers.home-assistant.io/docs/config_entries_config_flow_handler

MartinHjelmare commented on homeassistant/components/eq3btsmart/climate.py:
```diff
@@ -17,141 +28,441 @@
+DEVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_MAC): cv.string,
+        vol.Optional(CONF_SENSOR): cv.entity_id,
```
We don't allow accessing entities from other integrations in device integrating integrations. That kind of access is reserved to template integrations, "generic" integrations etc.

MartinHjelmare commented:
Since the premise for this change isn't compatible with our design architecture standards, I'll close the PR now. Thanks for your contribution!

andrewshadura commented:
@MartinHjelmar, but this PR is also about bugfixes and refactoring. Would you accept a version without config changes?

MartinHjelmare commented:
Yes, definitely. Please open a new PR for this. Thanks!

andrewshadura commented:
I’ve already made individual changes more atomic than they were before; I guess I need to work with @clau-bucur to go further :slightly_smiling_face: since he initially wrote those.

clau-bucur commented:
If you have the time and want you can take over, as I don't have too much currently. But we can discuss on Gitter.
